File: Go/Spoonerizer.go
package main

import (
	"fmt"
	"bufio"
	"os"
	"strings"
	"math/rand"
	"time"
)

func isVowel(c byte) bool {
	// Returns true if character is a vowel,
	// false otherwise.
	if c=='a'||c=='e'||c=='i'||c=='o'||c=='u' {
		return true;
	}
	return false;
}

func recombine(heads []string, tails []string) string {
	// Randomly combines the heads with the tails, and returns
	// result as a string.
	rand.Seed(time.Now().UTC().UnixNano());
	headIndex := int32(1 + rand.Intn(len(heads)-1));
	var output string;
	
	for i := 0; i < len(tails); i++ {
		if heads[headIndex] != "" || tails[i] != "" {
			output += heads[headIndex] + tails[i] + " ";
			headIndex = (headIndex + 1) % int32(len(heads));
		}
	}
	output = output[:len(output)-1];
	return output;
}

func splitInput(input []string) ([]string, []string) {
	// Splits each word from the input list into two lists:
	// heads = the characters up to the first vowel in the word.
	// tails = the characters from the first vowel to the end of the word.
	var heads []string;
	var tails []string;
	vowelIndex := 0;
	vowelFound := false;
	
	for i := 0; i < len(input); i++ {
		vowelIndex = 0;
		vowelFound = false;
		for j := 0; j < len(input[i]); j++ {
			if isVowel(input[i][j]) {
				vowelFound = true;
				vowelIndex = j;
				heads = append(heads, input[i][:j]);
				tails = append(tails, input[i][j:]);
				break;
			}
		}
		if !vowelFound {
			heads = append(heads, input[i]);
			tails = append(tails, "");
		}
	}
	return heads, tails;
}

func tokenize(s string) []string {
	// Custom tokenizing function.
	// Tokenizes on spaces.
	var output []string;
	var word []byte;
	spacesFound := 0;
	
	for i := 0; i < len(s); i++ {
		if s[i] == byte(' ') {
			if spacesFound == 0 && len(word) > 0 {
				output = append(output, string(word));
				word = make([]byte, 0);
			} else {
				spacesFound++;
			}
		} else {
			spacesFound = 0;
			word = append(word, s[i]);
		}
	}
	if len(word) > 0 {
		output = append(output, string(word));
	}
	return output;
}

func main() {
	reader := bufio.NewReader(os.Stdin);
	fmt.Print("Enter some words: ");
	inputText, _ := reader.ReadString('\n');
	inputText = strings.ToLower(inputText);
	inputText = strings.Trim(inputText, "\n");
	input := tokenize(inputText);
	
	if len(input) == 0 {
		fmt.Println();
	} else if len(input) == 1 {
		fmt.Println(input[0]);
	} else {
		heads, tails := splitInput(input);
		output := recombine(heads, tails);
		fmt.Println(output);
	}
}

File: Haskell/Spoonerizer.hs
import System.Random
import Data.String
import Data.Char

-- Returns true if character is a vowel, false otherwise.
isVowel :: Char -> Bool
isVowel c
	| c == 'a' = True
	| c == 'e' = True
	| c == 'i' = True
	| c == 'o' = True
	| c == 'u' = True
	| otherwise = False

-- Returns a random number between "lower" and "higher" inclusive.
getRandom :: Int -> Int -> IO Int
getRandom lower higher = do
	r <- (randomIO :: IO Int)
	return (mod r (higher - lower) + lower)

-- Returns the index of the first vowel found, or the length
-- of the string if a vowel is not found.
getVowelIndex :: String -> Int
getVowelIndex s =
	let vowelIndexes = [min i (length(s) - 1) | i <- [0..length(s)-1], isVowel(s !! i)]
	in if null vowelIndexes
		then -1
		else minimum vowelIndexes
	
-- Return the consonants before the first vowel of a word.
getHead :: (String, Int) -> String
getHead (s,i)
	| i < 0		= s
	| i == 0	= "" 
	| otherwise = take i s
	
-- Return the letters including and after the first vowel of a word.
getTail :: (String, Int) -> String
getTail (s, i)
	| i < 0		= ""
	| i == 0	= s
	| otherwise = drop i s
	
-- Returns a tuple of two lists:
-- 1. a list of the strings of consonants of each word that precede the first vowel
-- 2. a list of the strings of characters after and including the vowel, for each word
splitList :: [String] -> ([String], [String])
splitList input =
	let vowelIndexes = map getVowelIndex input;
		z = zip input vowelIndexes;
		heads = map getHead z;
		tails = map getTail z;
	in (heads, tails)

-- Recombine heads onto tails in a random order.
recombine :: ([String], [String]) -> Int -> String
recombine (h, t) r =
	let lenh = length h;
		h1 = take r h;
		h2 = drop r h;
		h3 = h2 ++ h1;
		rList = zip h3 t
	in unwords [x ++ y | (x, y) <- rList]

main :: IO ()
main = do
	putStrLn "Enter some words: "
	input <- getLine
	let inputWords = words (map toLower input);
		inputLength = length(inputWords);
		headsAndTails = splitList inputWords
	iorand <- (getRandom 1 inputLength)
	let r = iorand
	putStrLn ("\n\t--  " ++ (recombine headsAndTails r))
